### DISCUSSION TITLE
Is it possible to integrate with an OIDC identity provider, which encrypts its JWT tokens?

### DISCUSSION THREAD

**Discussion by konstantin-krasheninnikov-ullink** [keycloak/keycloak#28061] [Q&A] [answered]
We would like to integrate with an OIDC identity provider, which encrypts its JWS tokens ([RFC7516](https://www.rfc-editor.org/rfc/rfc7516)). There is a token decrypting endpoint available (security.oauth2.resource.token-info-uri).

Is there a way to integrate such a provider?

**Comment by justin-tay** [comment, marked as answer]
Keycloak does support brokering with an OIDC identity provider that encrypts its id tokens. The OIDC identity provider will need to be configured to use the public encryption keys configured in Keycloak, for instance by configuring the IDP with Keycloak's JWKS endpoint.

- https://github.com/keycloak/keycloak/pull/21355

That said I'm not sure what you mean by a token decrypting endpoint, that looks like a token introspection endpoint.

**Comment by konstantin-krasheninnikov-ullink** [comment]
I was hoping that I may have missed a way to give Keycloak the token introspection endpoint and disable the (broker) nonce checking for everything to work out of the box (so to speak). For now I am getting this exception:

`ERROR [org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider] Failed to make identity provider oauth callback: org.keycloak.broker.provider.IdentityBrokerException: OpenID Provider [oidc] did not return a nonce
        at org.keycloak.broker.oidc.OIDCIdentityProvider.preprocessFederatedIdentity(OIDCIdentityProvider.java:852)
        at org.keycloak.services.resources.IdentityBrokerService.authenticated(IdentityBrokerService.java:519)` 

That said I'll have to see if that OIDC identity provider can be configured to use the public encryption keys configured in Keycloak.

Thank you.

**Comment by justin-tay** [comment]
I think since Keycloak 23 there is an option to disable Keycloak from sending and verifying the nonce in the `id_token`.

Having the `id_token` encrypted won't change this behavior, it will still attempt to verify the nonce after decrypting the `id_token` if the checking is enabled.

![nonce](https://github.com/keycloak/keycloak/assets/49700559/33105407-2f4e-489d-a1f8-dff0a114637d)

**Comment by konstantin-krasheninnikov-ullink** [comment]
Hmm... if the IDP uses public keys from Keycloak's JWKS then it is, naturally, easy for Keycloak to enc./dec. the tokens using its private keys. But... isn't it natural for IDP to use private keys to encrypt its own tokens? Otherwise how, for example, can it serve data on a token introspection endpoint?

Is there any specification OIDC IDPs have to follow to know when to be a client (using public keys) or a server (using private keys)?

Basically, I would like to learn more about how exactly this "Support for JWE encrypted ID Tokens and UserInfo responses in OpenID Connect providers" should work in more details.

**Reply by justin-tay** [reply]
Public keys are used to encrypt and private keys to decrypt. It's not natural for an IDP to use it's own public keys to encrypt its own tokens since the only recipient that can decrypt it would be itself. If this is done it would be some internal implementation detail, for instance to have some sort of stateless operation and not part of any specification. And if this is really desirable it would likely be better for the IDP to just use a symmetric key to encrypt and decrypt the data.

The token introspection endpoint is intended for resource servers to get information about (typically opaque) access tokens, or for more information that's not in the claims if it's a JWT access token. This is not really related to OIDC and is more just OAuth2. Unlike access tokens, which can either be opaque or a JWT token, ID tokens must be a JWT token. OIDC specifies a `userinfo_endpoint` where you can receive information about a user by using the access token.

The "Support for JWE encrypted ID Tokens and UserInfo responses in OpenID Connect providers" is working as is specified. The tokens and responses are encrypted so that only the intended recipient, ie the client, can decrypt the information. The client public keys are used to encrypt the information. The client will then use it's private decryption keys to decrypt.

**Reply by konstantin-krasheninnikov-ullink** [reply]
Let's start from scratch if you allow...
- There's an OIDC IDP called, let's say, "Realcow." Realcow uses its private key to sign ID tokens and access tokens and encrypt them too because reasons.
- There's also (resource or web) server "Grassfield", which has a trust relationship with Realcow.
- Cow Fruh-fruh wants to use Grassfield and authenticates with Realcow via a redirect from Grassfield.
- Realcow does not have JWKS endpoint.
- Grassfield uses Realcow's token introspection endpoint to get information from an encrypted token because it can't decrypt it.
... 
And now Keycloak comes into play because another company uses it and wants Grassfield users to access their service called "Stalls&Breakfast." So, either custom user federation (Realcow is neither Kerberos not LDAP)  or OIDC IPD federation should work, except the Realcow's token is encrypted and Keycloak has to read it...

Well, I can imagine that if cow Fruh-fruh now goes to Keycloak and wants to use Realcow to log in, Keycloak redirects Fruh-fruh to Realcow, and after Fruh-fruh is authenticated and redirected back to Keycloak, Keycloak get the token encrypted by Realcow's private key and can use Realcow's public key (available somehow) to decrypt the token. But...

... in what specification do they recommend that IDP (Realcow in our example) should use a public key of an identity broker (like Keycloak) to encrypt its ID tokens? Is there some OIDC identity provider federation (or any other) spec, which has such recommendation? Or is it something Keycloak recommends?

That's what I was asking.  Anyway, if I understood your reply correctly...

It seem that Realcow has to recognize the redirect from Keycloak (and not from Grassfield) and, instead of using its private key (as usual), it has to go to Keycloak's JWKS endpoint, take the use: "enc" key from the set, encrypt the ID token and POST it back to Keycloak's redirect URL. Is it the workflow you were referring to?

**Reply by justin-tay** [reply]
It's a rather basic requirement to use the public key of the recipient for encryption otherwise the recipient then can't decrypt the data. Is this "Realcow" OIDC IDP a product or something custom developed?

In the [Example JWE](https://datatracker.ietf.org/doc/html/rfc7516#section-3.3) section of [RFC7516 JSON Web Encryption (JWE)](https://datatracker.ietf.org/doc/html/rfc7516#section-3.3)
`Encrypt the CEK with the recipient's public key using the RSAES-
      OAEP algorithm to produce the JWE Encrypted Key.`

In the [Key Encryption](https://datatracker.ietf.org/doc/html/rfc7516#appendix-A.1.3) section of [RFC7516 JSON Web Encryption (JWE)](https://datatracker.ietf.org/doc/html/rfc7516#appendix-A.1.3)
`Encrypt the CEK with the recipient's public key using the RSAES-OAEP
   algorithm to produce the JWE Encrypted Key.`

Your use case doesn't sound like identity federation. If "Grassfield" is a resource server and "Stalls&Breakfast" is an external resource server and each only accepts access tokens created by their respective authorization servers "Realcow" and "Keycloak" then actually if you don't need to propagate the user identity, "Grassfield" should use client credential flow to obtain an access token from "Keycloak" to call "Stalls&Breakfast" otherwise if the user identity is important, "Grassfield" would need to use token exchange to obtain the access token. Token exchange may be an issue for "Keycloak" as it's support for it is in preview and doesn't completely implement [RFC 8693 OAuth 2.0 Token Exchange](https://datatracker.ietf.org/doc/html/rfc8693).

**Comment by VisionWaves-config** [comment]
I am also facing the same kind of issue , I am configuring the Authetik endpoint as OIDC in keycloak. When users login I am getting private key not found in relam to decrypt the algorithm. Authtik team has shared the private key and certificate whihc I already added in key provider as RSA-OAEP-256 algorithm but no help. 

Any help will be apprecaited. Do I need to share the kid to Authentik team ?
